components/brave_wallet/browser/brave_wallet_pin_service.cc

```diff
@@ -85,14 +85,37 @@ absl::optional<mojom::WalletPinServiceErrorCode> StringToErrorCode(
   return absl::nullopt;
 }
 
-absl::optional<std::string> ExtractCID(const std::string& ipfs_url) {
-  GURL gurl = GURL(ipfs_url);
+absl::optional<std::string> ExtractIpfsUrl(const std::string& url) {
+  GURL gurl = GURL(url);
+
+  if (!gurl.is_valid()) {
+    return absl::nullopt;
+  }
+
+  if (gurl.SchemeIs(ipfs::kIPFSScheme)) {
+    return gurl.spec();
+  }
+
+  auto scope = ipfs::ExtractSourceFromGateway(gurl);
+  if (!scope || !scope->SchemeIs(ipfs::kIPFSScheme)) {
+    return absl::nullopt;
+  }
+
+  return scope.value().spec();
+}
+
+absl::optional<std::string> ExtractCID(const std::string& url) {
+  GURL gurl = GURL(ExtractIpfsUrl(url).value_or(""));
 
   if (!gurl.is_valid()) {
     return absl::nullopt;
   }
 
   if (!gurl.SchemeIs(ipfs::kIPFSScheme)) {
+    auto scope = ipfs::ExtractSourceFromGateway(gurl);
+    if (!scope || !scope->SchemeIs(ipfs::kIPFSScheme)) {
+      return absl::nullopt;
+    }
     return absl::nullopt;
   }
```

Comment:
I'm not sure if this block is still needed, if I understand correctly, when a valid GURL can be created from the return value of `ExtractIpfsUrl`, it will always be kIPFSScheme, no?

Reply:
Maybe just leave a DCHECK?